Tests/StoreKitUnitTests/ProductsManagerTests.swift

```diff
@@ -126,3 +128,75 @@ class ProductsManagerTests: StoreKitConfigTestCase {
     }
 
 }
+
+@available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+class ProductsManagerDiagnosticsTrackingTests: ProductsManagerTests {
+
+    private var mockDiagnosticsTracker: MockDiagnosticsTracker!
+
+    private var productsManager: ProductsManager!
+
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+
+        try AvailabilityChecks.iOS15APIAvailableOrSkipTest()
+
+        self.mockDiagnosticsTracker = MockDiagnosticsTracker()
+    }
+
+    func testFetchProductsWithIdentifiersSK1TracksCorrectly() throws {
+        let manager = self.createManager(storeKitVersion: .storeKit1,
+                                         diagnosticsTracker: self.mockDiagnosticsTracker)
+
+        let identifier = "com.revenuecat.monthly_4.99.1_week_intro"
+        _ = waitUntilValue(timeout: Self.requestDispatchTimeout) { completed in
+            manager.products(withIdentifiers: Set([identifier]), completion: completed)
+        }
+
+        expect(self.mockDiagnosticsTracker.trackedProductsRequestParams.count) == 1
+        let params = self.mockDiagnosticsTracker.trackedProductsRequestParams.first
+        expect(params?.wasSuccessful) == true
+        expect(params?.storeKitVersion) == .storeKit1
+        expect(params?.errorMessage).to(beNil())
+        expect(params?.errorCode).to(beNil())
+    }
+
+    func testFetchProductsWithIdentifiersSK2TracksCorrectly() throws {
+        let manager = self.createManager(storeKitVersion: .storeKit2,
+                                         diagnosticsTracker: self.mockDiagnosticsTracker)
+
+        let identifier = "com.revenuecat.monthly_4.99.1_week_intro"
+        _ = waitUntilValue(timeout: Self.requestDispatchTimeout) { completed in
+            manager.products(withIdentifiers: Set([identifier]), completion: completed)
+        }
+
+        expect(self.mockDiagnosticsTracker.trackedProductsRequestParams.count) == 1
+        let params = self.mockDiagnosticsTracker.trackedProductsRequestParams.first
+        expect(params?.wasSuccessful) == true
+        expect(params?.storeKitVersion) == .storeKit2
+        expect(params?.errorMessage).to(beNil())
+        expect(params?.errorCode).to(beNil())
+    }
+
+    #if swift(>=5.9)
+    @available(iOS 17.0, tvOS 17.0, macOS 14.0, watchOS 10.0, *)
+    func testFetchProductsWithIdentifiersSK2ErrorTracksCorrectly() async throws {
+        try AvailabilityChecks.iOS17APIAvailableOrSkipTest()
+
+        try await self.testSession.setSimulatedError(.generic(.unknown), forAPI: .loadProducts)
+        let manager = self.createManager(storeKitVersion: .storeKit2,
+                                         diagnosticsTracker: self.mockDiagnosticsTracker)
+
+        let identifier = "com.revenuecat.monthly_4.99.1_week_intro"
+        _ = try? await manager.products(withIdentifiers: Set([identifier]))
+
+        expect(self.mockDiagnosticsTracker.trackedProductsRequestParams.count) == 1
```

Comment:
one question... because I am facing the same issue. Since you're posting in a task inside the `ProductsManager` https://github.com/RevenueCat/purchases-ios/pull/4247/files#diff-b080ecaed1fab9b6af95a6a40653aa15720509e51c3faa1aa71263f99a4ff1cbR147 this check is going to be flaky, because you can't guarantee the task has executed, right?